### PR TITLE
Feature/validate CBnT image

### PR DESCRIPTION
**Pull request by ChriMarMe**
Validation of firmware image.

Checks KM and BPM structures and dependencies between.

### CODE REVIEW

**Comment by ChriMarMe** [comment]
I'll get a strange error when calling fit.GetTable(image) a second time.

> Test HasFitEntries: Valid  
Test KM Header: Valid  
Test KM Signature Offset: Valid  
Test KM Reserved2: Valid  
Test KM Revision: Valid  
Test KM SVN: Warning - KM SVN is zero. This indicates a development Security Version Number 
Test KM ID: Valid  
Test KM Pub Key Hash Alg: Valid  
2021/04/09 08:12:22 error parsing ME Flash Partition Table: ME Flash Partition Table signature 0x24465054 not found: first 20 bytes are:
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000010  00 30 00 00                                       |.0..|

**Comment by ChriMarMe** [comment]
Oh, I suspect the function does not work with coreboot images without the warning?

File: pkg/provisioning/bg/validation.go [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+		// NO BPM
+		return false, fmt.Errorf("No BPM entry in FIT found")
+	case 2:
+		// No KM
+		return false, fmt.Errorf("No KM entry in FIT found")
+	case 3:
+		// Success
+		return true, nil
+	default:
+		return false, fmt.Errorf("unexpected Error")
```
Comment (PatrickRudolph) [inline pkg/provisioning/bg/validation.go:201]: Add a test if there are two keymanifestrecords or two bootpolicymanifests as well?

File: pkg/provisioning/bg/validation.go [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+		s.WriteString("KM Header ID incorrect. Have: " + km.StructInfo.ID.String() + " - Want: " + string(headerID) + "\n")
+	}
+	if km.StructInfo.Version != headerVersion {
+		s.WriteString("KM Header Version incorrect. Have: " + string(km.StructInfo.Version) + " - Want: " + string(headerVersion) + "\n")
+	}
+	if km.StructInfo.Variable0 != 0 {
+		s.WriteString("KM Header Variable0 incorrect. Have: " + string(km.StructInfo.Variable0) + " - Want: 0")
+	}
+	if s.Len() != 0 {
+		return false, fmt.Errorf("%s", s.String())
```
Comment (PatrickRudolph) [inline pkg/provisioning/bg/validation.go:220]: Errors should not contain \n

File: pkg/provisioning/bg/validation.go [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+
+// KMStructureValidKMSOffset defines the behavior for the CBnTTest "KM Signature Offset"
+func KMStructureValidKMSOffset() (bool, error) {
+	var KMSigOffset uint16
+	//Count the bytes to Key Manifest Signature - Everything is static with the exception KeyHash and Key Manifest Signature
+	KMSigOffset = 24
+	// Count KMHashes
+	for _, hash := range km.Hash {
+		// Usage field 8 bytes static
+		KMSigOffset += 8
```
Comment (PatrickRudolph) [inline pkg/provisioning/bg/validation.go:233]: can you use binary.Size(some struct) here?

Comment (ChriMarMe) [reply pkg/provisioning/bg/validation.go:233]: When using binary.Size(hash) in the loop, I get -1 (which is not correct obviously) 
Calling binary.Size(km.Hash) also gets me -1. It seems go doesn't consider the structs as fixed length.

File: pkg/provisioning/bg/validation.go (hunk 1 above)
Comment (ChriMarMe) [reply pkg/provisioning/bg/validation.go]: Multiple KMs BPMs are not supported in general. Good idea for an issue.

**Comment by xaionaro** [comment]
> Any idea @rihter007 @xaionaro ?

Could you share the command you use?

**Comment by ChriMarMe** [comment]
/cbnt-prov validate coreboot.rom bpm_pub_key.pem km_pub_key.pem

File: cmd/cbnt-prov/cmd.go [hunk 4]
```diff
@@ -188,6 +188,13 @@ type keygenCmd struct {
 	Path     string `flag optional name:"path" help:"Path to store keys. File names are 'yourname_bpm/yourname_bpm.pub' and 'yourname_km/yourname_km.pub' respectivly"`
 }
 
+type validateCmd struct {
+	Path       string `arg required name:"path" help:"Path to firmware image under test"`
+	BpmPubKey  string `arg required name:"bpmpubkey" help:"Path to the Boot Policy Manifest public key file." type:"path"`
+	KmPubKey   string `arg required name:"kmpubkey" help:"Path to the Key Manifest public key file." type:"path"`
+	Interactiv bool   `flag optional name:"interactive" help:"Set this flag to pause after each error"`
```
Comment (walterchris) [inline cmd/cbnt-prov/cmd.go:195]: Interactive?

File: pkg/provisioning/cbnt/validation.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	bpm       *bootpolicy.Manifest
+}
+
+var (
+	image         []byte
+	bpmpubkeypath string
+	objUnderTest  testObj
+)
+
+func setObjUnderTest(imgpath, kmkeypath, bpmkeypath string) error {
```
Comment (walterchris) [inline pkg/provisioning/cbnt/validation.go:28]: maybe it's just me, but I like to have my arguments explicitly typed like imgpath string, kmkeypath string.. and so forth.

Comment (walterchris) [inline pkg/provisioning/cbnt/validation.go:28]: Also I would group that function beneath the const and vars.

File: pkg/provisioning/cbnt/validation.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	var err error
+	switch r {
+	case resultError:
+		_, err = s.WriteString("Error")
+	case resultNotRun:
+		_, err = s.WriteString("Not run!")
+	case resultValid:
+		_, err = s.WriteString("Valid")
+	case resultWarning:
+		_, err = s.WriteString("Warning")
```
Comment (walterchris) [inline pkg/provisioning/cbnt/validation.go:102]: Have a default case?

File: pkg/provisioning/cbnt/validation_help.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/9elements/converged-security-suite/v2/pkg/intel/metadata/manifest"
+	"github.com/tjfoc/gmsm/sm2"
+)
+
+func validateSignature(s *manifest.Signature, unsigneddata []byte, key crypto.PublicKey) error {
+	switch s.SigScheme {
+	case manifest.AlgRSAPSS:
```
Comment (walterchris) [inline pkg/provisioning/cbnt/validation_help.go:15]: nit: Both RSA cases could be grouped together?

**Review by walterchris — COMMENTED** [review]
Halfway through :)

File: cmd/cbnt-prov/cmd.go (hunk 4 above)
Comment (ChriMarMe) [reply cmd/cbnt-prov/cmd.go]: done

File: pkg/provisioning/cbnt/validation.go (hunk 5 above)
Comment (ChriMarMe) [reply pkg/provisioning/cbnt/validation.go]: done

File: pkg/provisioning/cbnt/validation.go (hunk 6 above)
Comment (ChriMarMe) [reply pkg/provisioning/cbnt/validation.go:57]: done

File: pkg/provisioning/cbnt/validation_help.go (hunk 7 above)
Comment (ChriMarMe) [reply pkg/provisioning/cbnt/validation_help.go]: done

File: pkg/provisioning/cbnt/validation.go (hunk 5 above)
Comment (ChriMarMe) [reply pkg/provisioning/cbnt/validation.go]: done

**Comment by ChriMarMe** [comment]
This solves #174

**Pull request closed without merging**